documentation/optee_design.md

```diff
@@ -327,69 +169,62 @@ Different stacks are used during different stages. The stacks are:
+used temporarily while assigning an appropriate stack pointer for `SP_EL0`.
+**`SP_EL1` is always assigned the value of `thread_core_local[cpu_id]`.** This
+structure have some spare space for temporary storage of registers and also
+keeps the relevant stack pointers. In general when we talk about assigning a
+stack pointer to the CPU below we mean `SP_EL0`.
 
 ## Boot
-During early boot the CPU is configured with the temp stack which is used
-until OP-TEE exits to normal world the first time.
+During early boot the CPU is configured with the temp stack which is used until
+OP-TEE exits to normal world the first time.
 
 *Notes for AArch64:*
-`SPSEL` is always `0` on entry/exit to have `SP_EL0` acting as stack
-pointer.
+`SPSEL` is always `0` on entry/exit to have `SP_EL0` acting as stack pointer.
 
 ## Normal entry
-Each time OP-TEE is entered from normal world the temp stack is used
-as the initial stack. For fast calls this is the only stack used. For
-normal calls an empty thread slot is selected and the CPU switches to
-that stack.
+Each time OP-TEE is entered from normal world the temp stack is used as the
+initial stack. For fast calls this is the only stack used. For normal calls an
+empty thread slot is selected and the CPU switches to that stack.
 
 ## Normal exit
-Normal exit occurs when a thread has finished its task and the thread is
-freed. When the main thread function, tee_entry_std(), returns interrupts
-are disabled and the CPU switches to the temp stack instead. The thread
-is freed and OP-TEE exits to normal world.
+Normal exit occurs when a thread has finished its task and the thread is freed.
+When the main thread function, tee_entry_std(), returns interrupts are disabled
+and the CPU switches to the temp stack instead. The thread is freed and OP-TEE
+exits to normal world.
 
 ## RPC exit
 RPC exit occurs when OP-TEE need some service from normal world. RPC can
-currently only be performed with a thread is in running state. RPC is
-initiated with a call to thread_rpc() which saves the state in a way that
-when the thread is restored it will continue at the next instruction as if
-this function did a normal return. CPU switches to use the temp stack
-before returning to normal world.
+currently only be performed with a thread is in running state. RPC is initiated
+with a call to thread_rpc() which saves the state in a way that when the thread
+is restored it will continue at the next instruction as if this function did a
+normal return. CPU switches to use the temp stack before returning to normal
+world.
 
 ## IRQ exit
-IRQ exit occurs when OP-TEE receives an IRQ, which is always handled in
-normal world. IRQ exit is similar to RPC exit but it's thread_irq_handler()
-that saves the thread state instead. The thread is resumed in the same way
-though.
+IRQ exit occurs when OP-TEE receives an IRQ, which is always handled in normal
+world. IRQ exit is similar to RPC exit but it's thread_irq_handler() that saves
```

Comment:
thread_irq_handler() and elx_irq() (respectively for ARMv7/Aarch32 and for Aarch64)